chore: Add a DA service unit test

EmandM opened the pull request:
<!-- Replace this block with what this PR does and why. Describe what you'd like reviewers to know, how you applied the engineering principles, and any interesting tradeoffs made. Delete bullet points below that don't apply, and update the changelog section as appropriate. -->

<!-- Add short version of the JIRA ticket to the PR title (e.g. "feat: new shiny feature [MTT-123]") -->
[MTT-11550](https://jira.unity3d.com/browse/MTT-11550)

## Changelog

- Added: A unit test against the da service 

## Testing and Documentation

- Includes unit tests.

<!--  Uncomment and mark items off with a * if this PR deprecates any API:
### Deprecated API
- [ ] An `[Obsolete]` attribute was added along with a `(RemovedAfter yyyy-mm-dd)` entry.
- [ ] An [api updater] was added.
- [ ] Deprecation of the API is explained in the CHANGELOG.
- [ ] The users can understand why this API was removed and what they should use instead.
-->

## Backport

<!-- If this is a backport:
 - Add the following to the PR title: "\[Backport\] ..." .
 - Link to the original PR.
If this needs a backport - state this here
If a backport is not needed please provide the reason why.
If the "Backports" section is not present it will lead to a CI test failure. 
-->

No backport applicable - DA only

EmandM committed:
chore: Add a DA service unit test

EmandM committed:
temp trigger update for faster iteration

EmandM committed:
fix pvp exceptions

EmandM committed:
Fix code docs

EmandM committed:
fix formatting

EmandM committed:
remove backticks

EmandM committed:
Add missing import

EmandM committed:
Swap the tests to run on ubuntu

EmandM committed:
use inline command instead of funtion

EmandM committed:
Fix while loop issue

EmandM committed:
Stop changeownership message on spawn

EmandM committed:
try connect to service utp only

EmandM committed:
update standalone service timeout to 1h

EmandM committed:
ensure active scene hash is always synchroniszed to the service

EmandM committed:
Add timeout between rust tests

EmandM committed:
unbreak test

EmandM committed:
Merge branch 'develop-2.0.0' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into chore/rust-unit-test

EmandM committed:
actually use the hosted service

EmandM committed:
Remove can connect test, we can connect. Run only CreateObjectNew()

EmandM committed:
Fix formatting

EmandM committed:
lets go

EmandM committed:
All tests passing

EmandM committed:
Run tests against release rust build

EmandM committed:
Fix InScenePlacedNetworkObjectTests

EmandM committed:
Only log cmb error logs

EmandM committed:
remove log lines

EmandM committed:
update port env variable

EmandM committed:
remove connection test

EmandM committed:
fix bug in StartServerAndClientsWithTimeTravel

EmandM committed:
cleanup

EmandM committed:
Merge branch 'develop-2.0.0' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into chore/rust-unit-test

EmandM committed:
further cleanup

NoelStephensUnity commented on com.unity.netcode.gameobjects/TestHelpers/Runtime/NetcodeIntegrationTest.cs:
```diff
@@ -127,8 +138,47 @@ public enum HostOrServer
+        protected NetworkManager[] m_NetworkManagers;
+
+        /// <summary>
+        /// Gets the current authority.
```
Maybe this should say:
Gets the current authority of the network session.

NoelStephensUnity commented on com.unity.netcode.gameobjects/TestHelpers/Runtime/NetcodeIntegrationTest.cs:
```diff
@@ -349,6 +414,7 @@ public IEnumerator SetUp()
         {
             VerboseDebugLog.Clear();
             VerboseDebug($"Entering {nameof(SetUp)}");
+            m_NumberOfClients = NumberOfClients;
```
Did you move this from OneTimeSetup to Setup for a specific reason?

EmandM replied:
I did not! Thanks for catching that!

NoelStephensUnity commented on com.unity.netcode.gameobjects/TestHelpers/Runtime/NetcodeIntegrationTest.cs:
```diff
@@ -1830,9 +1902,29 @@ public NetcodeIntegrationTest(NetworkTopologyTypes networkTopologyType, HostOrSe
+            InitializeTestConfiguration(m_NetworkTopologyType, hostOrServer);
+        }
+
+        private void InitializeTestConfiguration(NetworkTopologyTypes networkTopologyType, HostOrServer? hostOrServer)
```
❤️

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/Runtime/DistributedAuthority/NetworkClientAndPlayerObjectTests.cs:
```diff
@@ -238,23 +228,14 @@ private bool ValidateAllPlayerObjects()
-                }
 
-                if (m_ServerNetworkManager.LocalClient.PlayerObject == null)
+            foreach (var client in m_NetworkManagers)
```
❤️

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/Runtime/DistributedAuthority/OwnershipPermissionsTests.cs:
```diff
@@ -64,23 +65,13 @@ private bool ValidateObjectSpawnedOnAllClients()
-            }
 
-            foreach (var client in m_ClientNetworkManagers)
+            foreach (var client in m_NetworkManagers)
```
❤️

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkObject/NetworkObjectOwnershipTests.cs:
```diff
@@ -147,58 +145,61 @@ public IEnumerator TestOwnershipCallbacks([Values] OwnershipChecks ownershipChec
-            var serverObject = m_ServerNetworkManager.SpawnManager.SpawnedObjects[ownershipNetworkObjectId];
-            var clientObject = m_ClientNetworkManagers[0].SpawnManager.SpawnedObjects[ownershipNetworkObjectId];
-            Assert.That(serverObject, Is.Not.Null);
+            var authorityObject = authority.SpawnManager.SpawnedObjects[ownershipNetworkObjectId];
```
Really like your clean up within NetcodeIntegrationTest and how that improves the test flow... :godmode:

NoelStephensUnity commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkVariable/NetworkVariableCollectionsTests.cs:
```diff
@@ -1257,15 +1257,7 @@ protected override void OnServerAndClientsCreated()
-            }
-
-            foreach (var client in m_ClientNetworkManagers)
+            foreach (var client in m_NetworkManagers)
```
❤️

NoelStephensUnity commented on pvpExceptions.json:
Nice!
❤️

NoelStephensUnity commented on testproject/Assets/Tests/Runtime/NetworkSceneManager/InScenePlacedNetworkObjectTests.cs:
```diff
@@ -18,22 +18,27 @@ namespace TestProject.RuntimeTests
     public class InScenePlacedNetworkObjectTests : IntegrationTestWithApproximation
     {
-        protected override int NumberOfClients => 3;
+        protected override int NumberOfClients => 2;
```
Just curious why you reduced the client count here?

EmandM replied:
This test was creating three clients but implemented `CanStartServerAndClients()` to block automatically starting anything. The test then saved the third client and removed it manually from `m_ClientNetworkManagers` before calling `StartServerAndClients()`. It would later start that saved third client as a late join client.

Introducing `m_NetworkManagers` broke this flow as `StartServerAndClients()` was failing because that third client (which was still tracked in `m_NetworkManagers`)  hadn't been started, only the clients in the manually updated `m_ClientNetworkManagers` were started.

I updated the test to only use two clients and then to use `CreateAndStartNewClient()` to handle late joining a client.

NoelStephensUnity replied:
Ahhh... I see... yeah... makes sense!
I most likely wrote that when `CreateAndStartNewClient` didn't do all of the fancy stuff it does now... 👍

NoelStephensUnity reviewed (approved):
Only 1 or 2 questions and a minor suggestion on wording for API documentation.
Otherwise, very nice job:
- Cleaning up various areas within the NetcodeIntegrationTest.
  - Improved the "readability" in many places.
  - Added a more standard way to obtain various NetworkManager instances.
- Cleaning up various integration tests
  - Improved the "readability" in many places.
  - Created a more uniform standard "flow" and naming.
- Adding UseCMBService to tests that still needed conversion along with comments.
- Removing m_EnableVerboseDebug to reduce over-all test runner log size.

🥇 💯

EmandM committed:
code review comments

NoelStephensUnity committed:
text - fix

This is an attempt to fix an issue where Mac OSX is frequently failing the InterpolationStopAndStartMotionTest. Based on the error, I am thinking (hard to tell since it doesn't replicate locally in-editor nor stand alone) that it could be due to a Mac VM running slower than expected (randomly) which could cause the network tick event to trigger more than once in a single frame.

EmandM commented on .yamato/desktop-standalone-tests.yml:
```diff
@@ -81,21 +81,17 @@ desktop_standalone_test_{{ project.name }}_{{ platform.name }}_{{ backend }}_{{
     # Set this to ensure the DA codec tests will fail if they cannot connect to the echo-server
     # The default is to ignore the codec tests if the echo-server fails to connect
     ENSURE_CODEC_TESTS: "true"
+    USE_CMB_SERVICE: "true"
```
It's inside the `UseCMBService()` function inside `NetcodeIntegrationTest.cs` ([here](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/blob/db50deffef1ce80f832a19c84fcbe5acf85a5d1c/com.unity.netcode.gameobjects/TestHelpers/Runtime/NetcodeIntegrationTest.cs#L207))

EmandM committed:
add comments to bash file and explicitly pass env variables to script

EmandM commented on Tools/CI/run_cmb_service.sh:
```diff
@@ -0,0 +1,19 @@
+# clone the latest rust repo
```
Done! Added a bunch of info at the top of the file. I hope it makes things clearer
